### PR TITLE
[2.0.x] HAL for STM32F4

### PR DESCRIPTION
**Pull request by ktand**
### Description

HAL for STM32F4 using the [STM32GENERIC](https://github.com/danieleff/STM32GENERIC) Arduino core by danieleff. Based on the current STM32F1 & STM32F7 HALs.

All base features including support for the following is working:

- `REPRAP_DISCOUNT_SMART_CONTROLLER`
- `REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER`
- `I2C_EEPROM`
- Emulated Eeprom
- Endstop interrupts
- SD Card

TMC2130 is currently not working due to name clashing with STM32 Arduino Core. Requires changes to the TMC2130Stepper library (see[ TMC2130Stepper PR #30](https://github.com/teemuatlut/TMC2130Stepper/pull/30)).

`BOARD_STM32F4` & `pins_STM32F4.h` provided as a generic definitions for STM32F4 based boards.

### CODE REVIEW

File: Marlin/src/HAL/HAL_STM32F4/HAL_Servo_STM32F4.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef HAL_SERVO_STM32F4_H
+#define HAL_SERVO_STM32F4_H
+
+#include <../../libraries/Servo/src/Servo.h>
```
Comment (thinkyhead) [inline Marlin/src/HAL/HAL_STM32F4/HAL_Servo_STM32F4.h]: Any way to simplify this include path?

File: platformio.ini [hunk 2]
```diff
@@ -242,6 +242,18 @@ lib_ldf_mode = 1
...
+[env:STM32F4]
+platform    = ststm32
+framework   = arduino
+board       = disco_f407vg
+build_flags = ${common.build_flags} -DUSE_STM32GENERIC -DMENU_USB_SERIAL -DMENU_SERIAL=SerialUSB
+lib_deps    = ${common.lib_deps}
+lib_ignore  = Adafruit NeoPixel, c1921b4, TMC2130Stepper
+src_filter  = ${common.default_src_filter}
+upload_protocol = stlink
+upload_port = COM6
```
Comment (thinkyhead) [inline platformio.ini]: These upload port settings are unlikely to work on most platforms.

**Comment by thinkyhead** [comment]
Overall, looks very good. Just wondering about the complicated include path (`<../../libraries/Servo/src/Servo.h>`) and the `upload_port` setting before it gets merged.

**Comment by p3p** [comment]
There is no need for all the platform name postfixes on filenames, I realise you are just following what looks like the standard but I haven't got round to cleaning up the filenames on the other platforms, prefixes are also questionable but some headers will need something to differentiate them from standard library headers files or PlatformIO may get confused.

**Comment by ktand** [comment]
I have updated the PR addressing the concerns regarding the Servo.h include
path and the upload_port and upload_protocol parameters in platformio.ini.

I agree with the comments regarding the filenames but I think the change
should be done for all HALs, perhaps in a separate PR.

**Comment by thinkyhead** [comment]
> I agree with the comments regarding the filenames

Yes, a comprehensive cleanup pass is definitely called for. We've been in the **elaboration** phase for a while, but now that things are starting to work well the time is near for a **refactoring** phase.

**Pull request merged**